inc/saiswitch.h

```diff
@@ -3084,6 +3084,45 @@ typedef enum _sai_switch_attr_t
      */
     SAI_SWITCH_ATTR_EXTENDED_PORT_STATE_CHANGE_NOTIFY,
 
+     * @brief Tam telemetry reporting byte size of chunk
+     *
+     * Defines the maximum number of bytes in a single report.
+     * The total number of bytes in a report should be as close as possible to this value.
+     * We can increase this value to reduce the number of sys calls.
+     * If the type of message is IPFIX, this value should not be less than 65535.
+     * Because we don't expect the IPFIX record to be fragmented.
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_AND_SET
+     * @default 65535
+     */
+    SAI_SWITCH_ATTR_TAM_REPORT_CHUNK_SIZE,
+
+    /**
+     * @brief Tam telemetry chunk count
+     *
+     * This value indicates how many chunks of reports that can be restored in the buffer.
+     * If the data structure is a ring buffer, the byte size of ring buffer is chunk count * chunk size.
+     * 1 means that there is only one chunk of report in the buffer.
+     * If the buffer is full, new incoming data will be dropped.
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_ONLY
+     * @default 1
+     */
+    SAI_SWITCH_ATTR_TAM_CHUNK_COUNT,
+
+    /**
+     * @brief Set TAM telemetry type config change event notification callback function passed to the adapter.
+     *
+     * Use sai_tam_tel_type_config_change_notification_fn as notification function.
+     *
+     * @type sai_pointer_t sai_tam_tel_type_config_change_notification_fn
+     * @flags CREATE_AND_SET
+     * @default NULL
+     */
+    SAI_SWITCH_ATTR_TAM_TEL_TYPE_CONFIG_CHANGE_NOTIFY,
```

Comment:
Can this be moved to TAM object ?

Reply:
It looks like all notify handlers have to be bound on the switch object. It's the limitation of SAI design.

Reply:
I don't know if its SAI design limitation, or just that all callback handlers are at switch level currently. @kcudnik please comment.

Reply:
Hi Bandaru, yes, in my understanding, this is basically how the notify handler works. The handler will be used by all objects in that type, hence it is added on upper layers, and the best place is the switch, as it will be setup during the initialization.

For example, fob handler will be used by all fdb entries, hence it is not tied to any fdb entry.

This is why the notifications are added on the switch.

Reply:
> I don't know if its SAI design limitation, or just that all callback handlers are at switch level currently. @kcudnik please comment.

Currently it's forced that all notifications be in switch, this will unify entire design, and it will help update pointers when switch is put into warm boot mode, currently we don't have specific notifications per object and we are not planning to do so, more explanation is like Riff said in above comment

Reply:
> Hi Bandaru, yes, in my understanding, this is basically how the notify handler works. The handler will be used by all objects in that type, hence it is added on upper layers, and the best place is the switch, as it will be setup during the initialization.

In this case, all objects (tam_tel_type etc) are all under the single TAM object. That is the container and not switch object. Its not an upper layer, but a container. The comparision with FDB is also incorrect. Please see below.

> 
> For example, fob handler will be used by all fdb entries, hence it is not tied to any fdb entry.

The FDB notification (callback registration) is for all entries and not for a specific entry. There is no object that holds all FDB entries and thus switch is the only way out. 

> 
> This is why the notifications are added on the switch.

Reply:
is this a problem that all tam objects will be put over a single notification pointer in the switch? you still need to explicitly create "tam" object using switch_id, so it does not matter how internally tam is organizes, it still can use singe swithc notification to push notify and identify object that is affected

Reply:
Actually, the call back is specific to the TAM object that is linked to all "streaming telemetry" objects. That TAM object is a single one. However, in a given SAI Adapter instance there may be more than one TAM object, each TAM object serving a different feature. Thus, it doesn't make sense to link the callback to the SWITCH which implies that it is applicable for all TAM objects.

Reply:
you can return object_id of that tam in notification, and act accordingly, we currently are not planning to decouple pointers from switch to single objects, we generate a lot of metadat which help us organize everygnig, including notification pointers assignments, which are coupled to the switch, at this point it will be up to user to decide differnt action  based on notificaion data

Reply:
If we have a restriction in the metadata and other aspects, I think it is OK. I will wait for the name edits that I suggested above, before approving this PR. This callback item is closed from my side.

Reply:
Those are not hard restrictions, but we're made for convince, and drifting away from current schema would cause a lot of refactoring

Reply:
Thats OK, I am not talking about hard restrictions alone. No need to change all the infra for this for now.
 
However, I request you to consider the possibility that the callbacks may - at sometime in future - may be per-object and plan any changes for this.  Thanks @kcudnik

Reply:
there is also another side why all pointers are in switch, during warm boot of the platform, user is calling create_switch, and it can pass new pointers to the switch to receive notifications, this is importnt, since after platform warm boot, new notifications wii most likely have different memory address than previous boot, so notifications would need to be updated. and this could be done in a single call during create switch, if you would have pointers in objects , you would need to go and find that object and assign new notification on it or you would have invalid pointer, or null, depends how platform handles pointers in sdk